fact gathering, mounts, fixes for single proc code and tests

bcoca opened the pull request:
Instead of 2 code paths, use dummy pool
Moved pool selection to function for easier unit test
updated integration and added unit test

##### ISSUE TYPE

- Bugfix Pull Request

bcoca committed:
Fact gathering fix 'no shm' branhc

Added 'fake' pool framework that still handles timeout
updated tests

Co-authored-by: Matt Clay <matt@mystile.com>

mattclay commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -27,6 +27,7 @@
 
 from multiprocessing import cpu_count
 from multiprocessing.pool import ThreadPool
+from multiprocessing.dummy.pool import Pool as DummyPool
```
```suggestion
from multiprocessing.dummy import Pool as DummyPool
```

mattclay commented on test/units/module_utils/facts/hardware/linux/test_get_pool.py:
```diff
@@ -17,7 +17,7 @@ def test_get_pool():
 
-    with mock.patch('multiprocessing.pool.ThreadPool', new=mock.MagicMock()) as TPool:
-        TPool.__new__ = mock.MagicMock(side_effect=PermissionError('pooof'))
+    with mock.patch('multiprocessing.pool.ThreadPool.__new__', new=mock.MagicMock()) as TPool:
+        TPool = mock.MagicMock(side_effect=PermissionError('pooof'))
```
The mock can simplified further:

```suggestion
    with mock.patch.object(multiprocessing.pool.ThreadPool, '__init__', side_effect=PermissionError('poof')):
```

mattclay commented:
@bcoca It turns out the built-in dummy pool also uses `/dev/shm`, even though it doesn't need to, since it's only using threads. Thankfully there is a fairly simple solution for that:

```python
import threading

from multiprocessing.pool import ThreadPool, Pool
from multiprocessing.context import BaseContext

class DummyContext(BaseContext):
    def Lock(self):
        return threading.Lock()


class DummyPool(ThreadPool):
    def __init__(self, processes=None, initializer=None, initargs=()):
        Pool.__init__(self, processes, initializer, initargs, context=DummyContext())
```

This `DummyPool` can be used instead of `multiprocessing.dummy.Pool`.

mattclay commented on lib/ansible/module_utils/common/process.py:
```diff
@@ -4,11 +4,36 @@
+        Pool.__init__(self, processes, initializer, initargs, context=DummyContext())
+
+
+def get_thread_pool(size, warn=None):
```
Until we have a use case for making this utility code, we should probably keep it within the fact gathering code that needs it.

bcoca replied:
if it were the single function, I would agree, but if we add this kind of required scafolding, it makes little sense to keep there.

mattclay replied:
If we end up needing it in other modules later, we can move it into module_utils. Even then, unless there's a need to make it part of our public API, it should be private.

mattclay commented on lib/ansible/module_utils/common/process.py:
```diff
@@ -4,11 +4,36 @@
+
+class DummyPool(ThreadPool):
+    def __init__(self, processes=None, initializer=None, initargs=()):
+        Pool.__init__(self, processes, initializer, initargs, context=DummyContext())
```
```suggestion
        Pool.__init__(self, processes, initializer, initargs, context=DummyContext())  # pylint: disable=non-parent-init-called
```

mattclay commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -612,66 +602,44 @@ def get_mount_facts(self):
-            # serial processing, just assing results
-            mounts.append(results[mount]['info'])
-        else:
             pool.close()  # done with spawing new workers, start gc
```
The pool needs to be closed outside the for loop:

```suggestion
        pool.close()  # done with spawning new workers, start gc
```

mattclay committed:
Use concurrent.futures instead of multiprocessing

This entirely avoids the need for fallback logic since the concurrent.futures thread pool does not depend on `/dev/shm`.

mattclay commented on lib/ansible/plugins/action/copy.py:
```diff
@@ -310,13 +310,11 @@ def _copy_file(self, source_full, source_rel, content, content_tempfile,
+            # So we ensure that the remote is readable by remote user (could be moved to copy module)
             if remote_path:
-                self._fixup_perms2((self._connection._shell.tmpdir, remote_path))
+                self._fixup_perms2((self._connection._shell.tmpdir, remote_path), execute=False)
```
It looks like these changes are unrelated to fact gathering. Were they supposed to be on another PR?

bcoca replied:
yep

bcoca committed:
revert

mattclay commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -21,28 +21,22 @@
-from ansible.module_utils.common.text.converters import to_text
+from ansible.module_utils._internal._concurrent import _futures
 from ansible.module_utils.common.locale import get_best_parsable_locale
+from ansible.module_utils.common.text.converters import to_text
```
The original import above can be kept instead of moving it here.

bcoca replied:
its easier to find/update stuff if they are ordered

mattclay commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -21,28 +21,22 @@
 from ansible.module_utils.facts.utils import get_file_content, get_file_lines, get_mount_size
 from ansible.module_utils.six import iteritems
 
+
```
This extra line isn't needed.

bcoca committed:
removed extra while line
